vmdb/app/models/mixins/process_tasks_mixin.rb

```diff
@@ -0,0 +1,123 @@
+    end
+
+    def invoke_tasks_local(options)
+      options[:invoke_by] = task_invoked_by(options)
+      args = task_arguments(options)
+
+      instances, tasks = validate_tasks(options)
+
+      instances.each_with_index do |instance, idx|
+        task = MiqTask.where(:id => tasks[idx]).first
+
+        if task && task.status == "Error"
+          task_audit_event(:failure, options, :target_id => instance.id, :message => task.message)
+          task.state_finished
+          next
+        end
+
+        invoke_task_local(task, instance, options, args)
+
+        msg = "#{instance.name}: '#{options[:task]}' initiated"
+        task_audit_event(:success, options, :target_id => instance.id, :message => msg)
+        task.update_status("Queued", "Ok", "Task has been queued") if task
+      end
+    end
+
+    # default: invoked by task, can be overridden
+    def task_invoked_by(_options)
+      :task
+    end
+
+    # default: only handles retirement, can be overridden
+    def task_arguments(options)
+      options[:task] == 'retire_now' ? [options[:userid]] : []
+    end
+
+    # default implementation, can be overridden
+    def invoke_task_local(task, instance, options, args)
+      cb = {
+        :class_name  => task.class.to_s,
+        :instance_id => task.id,
+        :method_name => :queue_callback,
+        :args        => ["Finished"]
+      } if task
+
+      MiqQueue.put(
+        :class_name   => name,
+        :instance_id  => instance.id,
+        :method_name  => options[:task],
+        :args         => args,
+        :miq_callback => cb
+      )
+    end
+
+    # Helper method for invoke_tasks, to determine the instances and the tasks associated
+    def validate_tasks(options)
+      tasks = []
+
+      instances = base_class.where(:id => options[:ids]).order("lower(name)").to_a
+      return instances, tasks unless options[:invoke_by] == :task # jobs will be used instead of tasks for feedback
+
+      instances.each do |instance|
+        # create a task for each instance
+        task = MiqTask.create(:name => "#{instance.name}: '#{options[:task]}'", :userid => options[:userid])
+        tasks.push(task.id)
+
+        validate_task(task, instance, options)
+      end
+      return instances, tasks
+    end
+
+    # default: validate retirement, can be overridden
+    def validate_task(task, instance, options)
+      return true unless options[:task] == "retire_now" && instance.retired?
+      task.error("#{instance.name} is already retired")
+      false
+    end
+
+    def task_audit_event(event, task_options, audit_options)
+      options =
+        {
+          :event        => task_options[:task],
+          :target_class => base_class.name,
+          :userid       => task_options[:userid],
+        }.merge(audit_options)
+      AuditEvent.send(event, options)
+    end
+
+    def exception_if_task_unknown(options)
```

Comment:
The method name makes it sound like it will return the exception class, not actually raise it.  Perhaps `assert_known_task` or `raise_on_unknown_task`?